lighthouse-core/lib/csp-evaluator.js

```diff
@@ -0,0 +1,147 @@
+ * @property {number} severity
+ * @property {string} directive The directive the finding applies to.
+ * @property {string|undefined} value Keyword if the finding applies to one.
+ */
+
+const log = require('lighthouse-logger');
+const i18n = require('../lib/i18n/i18n.js');
+const {
+  Parser,
+  Type,
+  Directive,
+  evaluateForFailure,
+  evaluateForSyntaxErrors,
+  evaluateForWarnings,
+} = require('../../third-party/csp-evaluator/optimized_binary.js');
+
+const UIStrings = {
+  missingBaseUri: 'Missing base-uri allows the injection of base tags. ' +
+    'They can be used to set the base URL for all relative (script) ' +
+    'URLs to an attacker controlled domain. ' +
+    'Can you set it to \'none\' or \'self\'?',
+  missingScriptSrc: 'script-src directive is missing. ' +
+    'This can allow the execution of unsafe scripts.',
+  missingObjectSrc: 'Consider setting object-src to \'none\' to prevent the injection of ' +
+    'plugins that execute JavaScript.',
+  strictDynamic: 'Host allowlists can frequently be bypassed. Consider using ' +
+    '\'strict-dynamic\' in combination with CSP nonces or hashes.',
+  unsafeInline: '\'unsafe-inline\' allows the execution of unsafe in-page scripts ' +
+    'and event handlers. Consider using CSP nonces or hashes to allow scripts individually.',
+  unsafeInlineFallback: 'Consider adding \'unsafe-inline\' (ignored by browsers supporting ' +
+    'nonces/hashes) to be backward compatible with older browsers.',
+  allowlistFallback: 'Consider adding https: and http: url schemes (ignored by browsers ' +
+    'supporting \'strict-dynamic\') to be backward compatible with older browsers.',
+  reportToOnly: 'This CSP policy only provides a reporting ' +
+    'destination via the \'report-to\' directive. ' +
+    'This directive is only supported in Chromium-based browsers so it is ' +
+    'recommended to also use a \'report-uri\' directive.',
+  reportingDestinationMissing: 'This CSP policy does not configure a reporting destination. ' +
+    'This makes it difficult to maintain the CSP policy over time and monitor for any breakages.',
+  nonceLength: 'Nonces should be at least 8 characters long and use the base64 charset.',
+  missingSemicolon: 'Did you forget the semicolon? ' +
+    '{keyword} seems to be a directive, not a keyword.',
+  unknownDirective: 'Unknown CSP directive.',
+  unknownKeyword: '{keyword} seems to be an invalid keyword.',
+  deprecatedReflectedXSS: 'reflected-xss is deprecated since CSP2. ' +
+    'Please, use the X-XSS-Protection header instead.',
+  deprecatedReferrer: 'referrer is deprecated since CSP2. ' +
+    'Please, use the Referrer-Policy header instead.',
+  deprecatedDisownOpener: 'disown-opener is deprecated since CSP3. ' +
+    'Please, use the Cross-Origin-Opener-Policy header instead.',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/** @type {Record<number, Record<string, LH.IcuMessage>|string|LH.IcuMessage>} */
```

Comment:
```suggestion
/** @type {Record<number, string|LH.IcuMessage|Record<string, LH.IcuMessage>>} */
```

found the other way around to be a bit confusing